src/lib/link.ts

```diff
@@ -375,6 +383,87 @@ export class Link {
     );
   }
 
+  // Returns the last height that this side knows of the other blockchain
+  public async lastKnownHeader(side: Side): Promise<number> {
+    const { src } = this.getEnds(side);
+    const client = await src.client.query.ibc.client.stateTm(src.clientID);
+    return client.latestHeight?.revisionHeight?.toNumber() ?? 0;
+  }
+
+  // this will update the client if needed and relay all provided packets from src -> dest
+  // if packets are all older than the last consensusHeight, then we don't update the client.
+  //
+  // Returns all the acks that are associated with the just submitted packets
+  public async relayPackets(
+    source: Side,
+    packets: readonly PacketWithMetadata[]
+  ): Promise<AckWithMetadata[]> {
+    const { src, dest } = this.getEnds(source);
+
+    // check if we need to update client at all
+    const maxPacketHeight = packets.reduce(
+      (acc, { height }) => Math.max(acc, height),
+      0
+    );
+    let headerHeight = await this.lastKnownHeader(otherSide(source));
+
+    if (headerHeight < maxPacketHeight + 1) {
+      const curHeight = (await src.client.latestHeader()).height;
+      if (curHeight < maxPacketHeight + 1) {
+        await src.client.waitOneBlock();
+      }
+      headerHeight = await this.updateClient(source);
+    }
+
+    const submit = packets.map(({ packet }) => packet);
+    const proofs = await Promise.all(
+      submit.map((packet) => src.client.getPacketProof(packet, headerHeight))
```

Comment:
Maybe cleaner and easier to understand if the two `.map`s are combined: `packets.map(({ packet }) => src.client.getPacketProof(packet, headerHeight))`

Reply:
I did that first, but I need to use submit in the next line, and figured why do the pre-processing 2 times?

Reply:
Ah, didn't see that. 👍